### PR TITLE
added chunk-wide water scanning

### PR DESCRIPTION
**Pull request by LadySerenaKitty**
With chunk-wide water scanning, we now examine all water blocks and
fire off propagation as needed.

### CODE REVIEW

**Comment by Cervator** [comment]
Thanks :-)

Sending this one @skaldarnar's way for review.

File: src/main/java/org/teraslogy/simpleliquids/block/entity/LiquidEntitySystem.java [hunk 1]
```diff
@@ -143,4 +148,44 @@ public void update(float delta) {
...
+                            testPos.set(blockPos.x, blockPos.y, blockPos.z);
+                            testPos.add(side.getVector3i());
+                            // we only do this if we have air next to our liquid
+                            if (worldProvider.getBlock(testPos) == air) {
+                                blockUpdate(new OnChangedBlock(blockPos, water, water), worldProvider.getBlock(blockPos).getEntity());
+                                break; // GET TO THE CHOPPAH!
+                            }
+                        }
+                        // ___.___
+                        //  c00D`=--/
```
Comment (skaldarnar) [inline src/main/java/org/teraslogy/simpleliquids/block/entity/LiquidEntitySystem.java:175]: ??

File: src/main/java/org/teraslogy/simpleliquids/block/entity/LiquidEntitySystem.java [hunk 2]
```diff
@@ -143,4 +148,44 @@ public void update(float delta) {
...
+    public void onChunkGenerated(OnChunkGenerated chunkGenerated, EntityRef entity) {
+        chunkWorker(chunkGenerated.getChunkPos(), entity);
+    }
+
+    @ReceiveEvent(priority = EventPriority.PRIORITY_NORMAL)
+    public void onChunkLoaded(OnChunkLoaded chunkLoaded, EntityRef entity) {
+        chunkWorker(chunkLoaded.getChunkPos(), entity);
+    }
+
+    public void chunkWorker(Vector3i chunkPos, EntityRef entity) {
```
Comment (skaldarnar) [inline src/main/java/org/teraslogy/simpleliquids/block/entity/LiquidEntitySystem.java]: I would name it something else, mabye `processChunk` or similar. With `chunkWorker` I expect some worker class to be returned, which is not the case here

**Comment by skaldarnar** [comment]
based on quick testing this makes the module even less performant ... I noticed smaller lags all the time, although the framerate dropped by only a few fps. Nothing critical for merging, though

File: src/main/java/org/teraslogy/simpleliquids/block/entity/LiquidEntitySystem.java (hunk 2 above)
Comment (LadySerenaKitty) [inline src/main/java/org/teraslogy/simpleliquids/block/entity/LiquidEntitySystem.java]: Sure, I can rename it, just give me a bit.

**Comment by LadySerenaKitty** [comment]
Theoretically, it should work with just onChunkGenerated.  I added onChunkLoaded just to make sure water gets filled out properly.

File: src/main/java/org/teraslogy/simpleliquids/block/entity/LiquidEntitySystem.java (hunk 1 above)
Comment (LadySerenaKitty) [inline src/main/java/org/teraslogy/simpleliquids/block/entity/LiquidEntitySystem.java:175]: The `break;` has the comment `GET TO THE CHOPPAH!`  So I added a helicopter to indicate where the instruction pointer lands upon executing the `break;`

Comment (msteiger) [inline src/main/java/org/teraslogy/simpleliquids/block/entity/LiquidEntitySystem.java:175]: LOL - this is hilarious!

**Comment by msteiger** [comment]
Similar code exists in various places, even in terrain generators themselves, so in theory this should be fairly fast. Maybe you can try with VisualVM to find the bottleneck.

**Comment by Cervator** [comment]
Bump - I was looking for poor lost PRs that may have been forgotten. I see one commit was added after the last comment, but that may not have triggered any notifications. This should!

Anybody want to double check on this and see if we should merge yet?

**Comment by iojw** [comment]
I've tested this out and it seems to work good. There are a few minor lag spikes but that's likely just my laptop since it happens with default worlds too .

These changes would allow propagation to apply to generated water blocks as well, which I feel would be a good thing to add. One thing I would like to include is a check for space below the current block since it currently only checks the sides. Other than that, I think that this PR looks good! I could merge it in a separate commit containing the above changes.

@Cervator @skaldarnar Any comments? 🙂

**Comment by Cervator** [comment]
@iojw: Thanks for the initiative to check out this module again after so long! I had this PR in the back of my mind for the longest of time, but no said time to actually get back to it.

I just tested it but on my new PC which is probably not particularly vulnerable to performance issues .. I was able to railgun myself into a brand new and very deep ocean hole / tunnel thing that filled nicely (although not completely) with water. Should probably test it with Caves enabled sometime and see how that works out, that should trigger some chunkgen style water propagation!

Go ahead and merge this, you should have access to do that now :-)

@LadySerenaKitty: hello again! Long time no see, hope you are doing well and thanks again for this PR :-)

**Comment by iojw** [comment]
Closing as merged in ea3fcea107c1a3b5e427452e1c1aed1903f0a53a along with an added check for the bottom block in a2ba53425eb0f274ec30dc054c2c7d7e4fa0fbf0

Thanks @LadySerenaKitty! 😄

File: src/main/java/org/teraslogy/simpleliquids/block/entity/LiquidEntitySystem.java (hunk 1 above)
Comment (VaibhavBajaj) [reply src/main/java/org/teraslogy/simpleliquids/block/entity/LiquidEntitySystem.java:175]: LOOOOOL! Never seen better

**Pull request closed without merging**